[#2468] R2DBC support

riba2101 opened the pull request:
This addes R2DBC support (pre-alpha) to QueryDsl.

Fixes #2468

jwgmeligmeyling commented:
Thanks @riba2101 , I'll have a look tomorrow into your compilation issues

riba2101 commented:
There are no compilation issues, r2dbc tests are failing, will take a look tomorrow

jwgmeligmeyling commented:
Ah sorry, I misread your commit message. I'll have look regardless 😄

riba2101 commented:
so...fixed locally everything -> could you pls provide a detailed report + tell me what versions you run so that i can do a similar setup locally

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchable.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Get the projection as a unique result.
+     * Consuming the result produces {@link NullPointerException} if the value is null.
+     * Consuming the result produces {@link IndexOutOfBoundsException} if there is more than one matching result
```
For consistency, this case should really throw `NonUniqueResultException`

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchable.java:
```diff
@@ -0,0 +1,46 @@
+
+    /**
+     * Get the projection as a unique result.
+     * Consuming the result produces {@link NullPointerException} if the value is null.
```
Instead of throwing a `NullPointerException`, you should return `Mono.empty()`

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchable.java:
```diff
@@ -0,0 +1,46 @@
+
+    /**
+     * Get the first result of the projection.
+     * Consuming the result produces {@link NullPointerException} if the value is null.
```
Instead of throwing a `NullPointerException`, you should return `Mono.empty()`

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchable.java:
```diff
@@ -0,0 +1,46 @@
+
+    /**
+     * Get the projection as a typed Flux.
+     * Consuming the result produces {@link NullPointerException} if a value is null.
```
Instead of throwing a `NullPointerException`, you should return `Flux.empty()`

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchableQuery.java:
```diff
@@ -0,0 +1,41 @@
+     * @param expr new projection
+     * @return the current object
+     */
+    <U> ReactiveFetchableQuery select(Expression<U> expr);
```
```suggestion
    <U> ReactiveFetchableQuery<U, ?> select(Expression<U> expr);
```

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/ReactiveFetchableQuery.java:
```diff
@@ -0,0 +1,41 @@
+     * @param exprs new projection
+     * @return the current object
+     */
+    ReactiveFetchableQuery select(Expression<?>... exprs);
```
```suggestion
    ReactiveFetchableQuery<Tuple, ?> select(Expression<?>... exprs);
```

jwgmeligmeyling commented on querydsl-core/src/main/java/com/querydsl/core/support/ReactiveFetchableQueryBase.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public Mono<T> fetchOne() {
+        return fetch().singleOrEmpty();
```
This actually implements the correct behaviour, make sure to update the Javadoc in the interface.

riba2101 replied:
yeah the Javadoc wasn't changed - did this

jwgmeligmeyling commented on querydsl-r2dbc/src/main/java/com/querydsl/r2dbc/H2Templates.java:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author tiwe
+ */
+public class H2Templates extends SQLTemplates {
```
H2 templates is on the classpath (you're depending on `querydsl-sql` ). Why is there a need to also ship it here?

I would maybe like to see the `H2RdbcTemplates` extend the actual `H2Templates`. At least then we do not have to worry about the two template variants diverging too much leading to bugs.

But is it really necessary? I am fine with extending the `SQLTemplates`  API in `querydsl-sql`  with a couple of methods that you might need for proper R2dbc support.

riba2101 replied:
the reason behind this is that the underlying super `SQLTemplates` differs because of a single line `com.querydsl.r2dbc.SQLTemplates#serializeInsert:919` (`R2DBCInsertBatch` vs `SQLInsertBatch`), yet again didn't know how invasive I'm allowed to be

jwgmeligmeyling commented on querydsl-r2dbc/src/main/java/com/querydsl/r2dbc/JDBCTypeMapping.java:
```diff
@@ -0,0 +1,174 @@
+ *
+ * @author tiwe
+ */
+final class JDBCTypeMapping {
```
Just use the one from `querydsl-dsl` . Make it `public` if you have to.

riba2101 replied:
see next, here we found that Blob/Clob has a special type for r2dbc - im 99% sure this will change when we change/map to correct r2dbc types (for example r2dbc heavily depends on java.time api)

jwgmeligmeyling commented on querydsl-r2dbc/src/main/java/com/querydsl/r2dbc/JavaTypeMapping.java:
```diff
@@ -0,0 +1,162 @@
+ *
+ * @author tiwe
+ */
+class JavaTypeMapping {
```
Just use the one from `querydsl-dsl` . Make it `public` if you have to.

riba2101 replied:
checked this -> so there are types for r2dbc that changed and this registry is for those types (notice the import `import com.querydsl.r2dbc.types.*;`

riba2101 commented:
Yeah, there are def Javadoc inconsistencies - sry for that, was kinda speed running through the code. 

I totally agree with you about the duplicate code problems(101%). Hate them, and it is a maintainability nightmare. I would love to see improvements in extensibility for such a case where the signature of a method differs from the base in the version 5. 

The problem was that I didn't know how invasive I can be :) 

This being said, there are only a couple of places where I could introduce other changes(will do them) without introducing massive changes to the querydsl-sql Or would need to refactor half of SQL to make it extensible. What's your take on this?

jwgmeligmeyling commented:
> . I would love to see improvements in extensibility for such a case where the signature of a method differs from the base in the version 5.

@johnktims are you fine with making some of the classes duplicated here `public`  instead?

johnktims commented:
Oh yeah. I’d much rather make the classes `public` than copy them.

mp911de commented on querydsl-r2dbc/pom.xml:
```diff
@@ -0,0 +1,226 @@
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
```
Nit: Please rather import `r2dbc-bom` for dependency management. This way, you don't require to specify individual driver versions.

riba2101 replied:
yeah, did it already locally -> this is a heavy WIP :)

riba2101 commented:
Hey guys,

I'm still at this but lately didn't have time at all.
I've kinda hit a hard wall -> R2DBC does not support java.util.Date/Time/java.sql.Date...(except Postgres) and to do that I would need to add to BeanPath the create methods for java.time API which comes in java 8, so would need to upgrade the java version which would be a major change? 

what now? halt this and wait for querydsl 5? do the java upgrade? get a beer? 

BR

mp911de commented:
R2DBC uses JSR-310 types. You should be able to wrap the `Result`/`Row` to return plain old dates. In a similar fashion, wrapping `Statement` to pass plain old dates as arguments. Your wrappers would then handle the conversion.

riba2101 commented:
If I do it like that querydsl would only support plain old dates, or we would have a double conversion (i did this). Then stumbled on the next problem(s):

1. the whole date support of querydsl is not supported, but mapping is working
2. Drivers don't have an ObjectCodec but rather throw a IllegalArgumentException - "codec not found"
3. Drivers don't support partial SQL dates, only the full format - for example if I say `YEAR(some_year_col)` and get back 1980, the driver(h2 for example) throws an exception that it can't parse a date because it expects the ISO format or falls back to the yyyyMMdd format(this is a problem with the driver)

mp911de commented:
Can you file a bug report against R2DBC H2? The next service release for r2dbc-h2 is scheduled for next week so we can include fixes until then.

riba2101 commented:
yeah ofc, wanted to do it tomorrow. unfortunately, it is not the only bug I've found and am collecting them (the r2dbc support is close to stable now).

riba2101 commented:
hi guys, 

so, added a lot of stuff, found various r2dbc (driver) bugs -> will open them shortly
there are a lot of changes i know, but for the sake of pulling this trough, added an example in r2dbc

known issue: posgress is swallowing inserts

jwgmeligmeyling commented:
Hi @riba2101, any new status updates? Anything where we can help?

riba2101 commented:
Hi @jwgmeligmeyling.

actually it is "done". There is an example spring project. PR needs to be checked and I need to open various R2DBC bugs.
For the time being Postgres support is disabled because some inserts get swallowed up.

Meaningful docs need to be written, but tbh I'm kinda flooded with work

jwgmeligmeyling reviewed (changes requested):
* Branch should be up to date with develop
* `maven-enforcer-plugin` should be reenabled.
   * This isn't going to cause issues, because the current develop targets a 5.x release which allows for breaking API changes.
* All the unnecesairy visibility changes should be undone, PR should be reduced to the bare minimum
* R2DBC dependencies should be removed from core, interfaces that rely on it should be moved to the r2dbc module.

riba2101 commented:
hmmm the tests are tunning for 5h already, no output, can someone check?

aoudiamoncef commented:
Hi @jwgmeligmeyling @riba2101,

IMHO it could be better to make  `querydsl-core` generic using [Reactive Streams](https://www.reactive-streams.org/) in order to have multiple [implementations ](https://en.wikipedia.org/wiki/Reactive_Streams) like Reactor, RxJava..etc

riba2101 commented:
this has been my point and a lot of ground work was already been done, thus the size of the pr, but then again, the version 5 requires another update and since this is already slow, i wanted to see the changes first in the version 5 and get a go ahead for such invasive changes.

would be also nice to get help since its already huge

aboutZZ commented:
Any updates of this PR?

riba2101 commented:
as already mentioned, im kinda super busy lately and would appreciate help!

chenzhenjia commented:
Is querydal no longer maintained?
